common/socks5/proxy-helpers/src/connection_controller.rs

```diff
@@ -127,21 +132,23 @@ impl Controller {
     fn send_to_connection(&mut self, conn_id: ConnectionId, payload: Vec<u8>, is_closed: bool) {
         if let Some(active_connection) = self.active_connections.get_mut(&conn_id) {
             if !payload.is_empty() {
-                active_connection.write_to_buf(payload);
+                active_connection.write_to_buf(payload, is_closed);
             } else if !is_closed {
                 error!("Tried to write an empty message to a not-closing connection. Please let us know if you see this message");
             }
-            // if messages get unordered, make sure we don't lose information about
-            // remote socket getting closed!
-            active_connection.is_closed |= is_closed;
 
             if let Some(payload) = active_connection.read_from_buf() {
+                if let Some(closed_at_index) = active_connection.closed_at_index {
+                    if payload.last_index > closed_at_index {
```

Comment:
is it possible for payload's `last_index` to be greater than `closed_at_index`? isn't it the case we expect the last received packet to be the closing packet?

Reply:
Yes. When reading the `OrderedBuffer`, the index returned actually indicates the number of messages returned so far (or the index of the one yet to come), kinda like TCP SEQ and ACK numbers. 
So if for example the closing message has index 30, the `OrderedBuffer` will deliver that message along with the index 31, effectively making `last_index` bigger than `closed_at_index` when we want to close the connection.  
I've proposed it that way to avoid messing with index's offset in the `OrderedBuffer::read()` method.